beps/bep_0003.rst

```diff
@@ -87,70 +92,153 @@ announce
 
 info
   This maps to a dictionary, with keys described below.
+  
+``piece layers``
+  A dictionary of strings. For each file in the file tree that is larger than the piece size
+  it contains one string value.
+  The keys are the merkle roots while the values consist of concatenated hashes
+  of one layer within that merkle tree.
+  The layer is chosen so that one hash covers ``piece length`` bytes.
+  For example if the piece size is 16KiB then the leaf hashes are used.
+  If a piece size of 128KiB is used then 3rd layer up from the leaf hashes is used.
+  Layer hashes which exclusively cover data beyond the end of file,
+  i.e. are only needed to balance the tree, are omitted.
+  All hashes are stored in their binary format.
+  
+  A torrent is not valid if this field is absent, the contained hashes do not match
+  the merkle roots or are not from the correct layer.
```

Comment:
> Torrents with both v1 and v2 pieces lists could get pretty bloated. 

That doubles the size. which can be compensated by doubling the piece size, more or less. Serving gzipped .torrents from disk with `Transfer-Encoding: gzip` (which torrent caches do to my knowledge) will also keep the duplicate filenames cheap. So that doesn't look like a strong argument to me.

> Some sites may want to reduce the size of the torrent file without going all the way to magnet links.

I'm not quite seeing the advantage. You don't get the file-reuse advantage of a full torrent anyway, so you need to wait for peers to do anything with it. So you might as well use a magnet.

> BEP 9 is already a de facto requirement due to the prevalence of magnet links so it doesn't seem like that much of a burden to depend on it.

Something about that reasoning seems incongruous to me. If you're already assuming BEP9 support then you can use magnets. torrents-without-layers only make sense in the narrow problem space where you don't have BEP9 support but want a lightweight transfer and have a core protocol that can transfer hashes. This is quite far away from my mental model of how things should work.

Either you go through a `.torrent` and are *no worse off than today* feature-wise. Or you go through magnets and you get faster startup than today.

In fact you will already be better off with the new torrent format since deduplication will be far far easier. And if server admins really really care about footprint they could use highly optimized storage to deduplicate the `piece layers` keys. Although I don't think they have a reason to, a 10M torrent archive should easily fit on a 1-2TB disk.

This covers both ends of the spectrum to the maximum extent possible. I don't see why we need to shift things towards one end. Basically, it seems like you want what I included in BEP9 in the core protocol. But I'd like to see a better argument why that's necessary and can't be covered by the proposal as-is.

I'm trying to get to the point where we can have (nearly) directory-layout oblivious torrent clients. For that those hashes are really needed. Right now scanning a directory tree for matching files is possible but tedious to implement due to the alignment issues. With aligned, fixed block size merkle trees *and the piece layers* it becomes possible to crawl large directory trees within seconds (your IOPS permitting) to locate files when they're missing e.g. because users moved or renamed them. This is important for endusers with large media libraries to keep torrents seeded without hassle. Seeding and keeping media organized are often somewhat at odds.